### PR TITLE
control/controlclient, types/netmap: start plumbing delta netmap updates

### PR DESCRIPTION
**Pull request by bradfitz**
Currently only the top four most popular changes: endpoints, DERP home, online, and LastSeen.

Updates #1909

### CODE REVIEW

File: control/controlclient/auto.go [hunk 1]
```diff
@@ -752,6 +776,25 @@ func (q *execQueue) Add(f func()) {
 	}
 }
 
+func (q *execQueue) RunSync(ctx context.Context, f func()) error {
+	for {
+		if err := q.wait(ctx); err != nil {
```
Comment (maisem) [inline control/controlclient/auto.go:781]: explain why?

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -1121,6 +1120,56 @@ func (b *LocalBackend) SetControlClientStatus(c controlclient.Client, st control
 	b.authReconfig()
 }
 
+var _ controlclient.NetmapDeltaUpdater = (*LocalBackend)(nil)
+
+// UpdateNetmapDelta implements controlclient.NetmapDeltaUpdater.
+func (b *LocalBackend) UpdateNetmapDelta(muts []netmap.NodeMutation) (handled bool) {
+	mc, err := b.magicConn()
+	if err != nil {
+		return false
```
Comment (maisem) [inline ipn/ipnlocal/local.go:1129]: log?

**Review by maisem — APPROVED** [review]

**Comment by maisem** [comment]
`TestMutationsFromMapResponse` seems to have an ordering problem based on https://github.com/tailscale/tailscale/actions/runs/6161095880/job/16719550258?pr=9244

```
    nodemut_test.go:195: wrong result (-want +got):
          []netmap.NodeMutation{
        - 	netmap.NodeMutationOnline{mutatingNodeID: 1, Online: true},
          	netmap.NodeMutationOnline{mutatingNodeID: 2},
        + 	netmap.NodeMutationOnline{mutatingNodeID: 1, Online: true},
          }
```

File: types/netmap/netmap.go [hunk 3]
```diff
@@ -124,6 +125,31 @@ func (nm *NetworkMap) PeerByTailscaleIP(ip netip.Addr) (peer tailcfg.NodeView, o
...
+// PeerIndexByNodeID returns the index of the peer with the given nodeID
+// in nm.Peers, or -1 if nm is nil or not found.
+//
+// It assumes nm.Peers is sorted by Node.ID.
+func (nm *NetworkMap) PeerIndexByNodeID(nodeID tailcfg.NodeID) int {
+	if nm == nil {
+		return -1
+	}
+	idx, ok := sort.Find(len(nm.Peers), func(i int) int {
+		v := nm.Peers[i].ID() - nodeID
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply types/netmap/netmap.go:137]: oh, right! forgot we had that.

File: types/netmap/nodemut.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Tailscale Inc & AUTHORS
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply types/netmap/nodemut.go:1]: The idea was there'd be other mutation stuff later.

File: wgengine/magicsock/magicsock.go [hunk 5]
```diff
@@ -2588,6 +2588,29 @@ func simpleDur(d time.Duration) time.Duration {
...
+		nodeID := m.NodeIDBeingMutated()
+		ep, ok := c.peerMap.endpointForNodeID(nodeID)
+		if !ok {
+			continue
+		}
+		switch m := m.(type) {
+		case netmap.NodeMutationDERPHome:
+			ep.setDERPHome(uint16(m.DERPRegion))
+		case netmap.NodeMutationEndpoints:
+			ep.mu.Lock()
```
Comment (maisem) [inline wgengine/magicsock/magicsock.go:2606]: Is the lock ordering documented

**Review by maisem — APPROVED** [review]

File: wgengine/magicsock/magicsock.go (hunk 5 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:2606]: yup:

```go
	// mu protects all following fields.
	mu sync.Mutex // Lock ordering: Conn.mu, then endpoint.mu
```

**Pull request merged**